feat(perps): add home screen v2, market sorting, and batch operations

abretonc7s opened the pull request:
## **Description**

This PR implements seven major improvements to the Perps trading interface, delivered as a cohesive enhancement to the user experience:

### 1. **Perps Home Screen v2** (TAT-1538)

Replaced the previous market-only list view with a comprehensive trading dashboard that provides an at-a-glance overview of all relevant perps information:

- **Positions Carousel**: Horizontal scrollable list of open positions with quick access to position details
- **Orders Carousel**: Horizontal scrollable list of open limit orders with quick navigation
- **Trending Markets List**: Top 5 markets by 24h trading volume with "See all" navigation
- **Recent Activity List**: Last 3 trade executions with "See All" link to full activity tab
- **Learn More Component**: Educational component for new users (moved from individual market pages)

### 2. **Market Sorting & Filtering** (TAT-1539)

Added comprehensive market sorting capabilities accessible via the "See all" button from trending markets:

- **Sort Fields**: Volume, Price Change, Open Interest, Funding Rate
- **Sort Direction**: Simplified list-based selection with combined field+direction options (e.g., "Price change: high to low")
- **Search**: Real-time market search by symbol/name
- **Watchlist Filter**: Optional toggle to show only favorited markets
- **Bottom Sheet Selection**: Native bottom sheet with checkmark indicators for selected options
- **Persistent Preferences**: User's sort field and direction preferences saved per network (testnet/mainnet) in controller state, applied automatically on next visit

#### Design Decision: 24h Timeframe Only

The initial design included a timeframe selector (1h / 24h) for Volume and Price Change data. However, **we've intentionally omitted the 1h timeframe** for the following technical reasons:

**API Limitation:**

- HyperLiquid's `metaAndAssetCtxs` API endpoint only provides 24h aggregated data:
  - `dayNtlVlm` (24h notional volume)
  - `prevDayPx` (previous day price for 24h change calculation)
- No equivalent 1h fields (`hourNtlVlm`, `prevHourPx`) exist in the API response

**Performance Concerns:**

- While HyperLiquid supports 1h candle subscriptions, implementing 1h metrics would require:
  - Subscribing to individual candle WebSocket feeds for **every market** (~100+ assets)
  - Extracting volume and price data from the latest completed 1h candle per asset
  - Maintaining ~100+ active WebSocket subscriptions just for market list sorting
  - Significantly increased API calls, memory usage, and complexity
- This approach would be **inefficient and not aligned with the current architecture**, which uses a single aggregated endpoint for all market data

**User Impact:**

- Market sorting functionality remains fully functional with 24h data
- All other metrics (Open Interest, Funding Rate) are point-in-time snapshots and don't require timeframe selection
- When/if HyperLiquid adds 1h aggregated data to their API, the timeframe selector can be easily restored (component already exists but is unused)

### 3. **Close All Positions** (TAT-1049)

Implemented batch position closing to streamline portfolio management:

- **Close All Button**: Visible on home screen when user has ≥2 open positions
- **Summary Bottom Sheet**: Displays aggregated margin, P&L, fees, and receive amount before confirmation
- **Shared Summary Component**: Extracted `PerpsCloseSummary` component used across single position close, close all modal, and close all view (eliminated 138 lines of duplicate JSX)
- **Batch Controller Method**: New `closePositions()` method in PerpsController with parallel execution and partial success handling
- **Enhanced Fee Hook**: Updated `usePerpsOrderFees` to accept single OR array params with automatic aggregation

### 4. **Cancel All Orders** (TAT-1815)

Implemented batch order cancellation following the same UX pattern as Close All Positions:

- **Cancel All Button**: Visible when user has multiple open orders
- **Confirmation Modal**: Simple confirmation dialog with order count
- **Batch Controller Method**: New `cancelOrders()` method in PerpsController with parallel execution
- **Consistent UX**: Follows same bottom sheet and confirmation pattern as Close All Positions

### 5. **Market Watchlist** (TAT-1252)

Added market watchlist functionality for quick access to preferred markets:

- **Watchlist Icon**: Star icon in market details header toggles watchlist status with immediate visual feedback
- **Persistent Storage**: Watchlist saved per network (testnet/mainnet) in controller state, survives app restarts
- **Activity Link Relocation**: Moved "Go to your activity" from header icon to Overview tab content for better UX
- **MetaMetrics Tracking**: Add to/remove from watchlist actions tracked with asset symbol and total watchlist count

### 6. **Saved Trade Configuration** (TAT-1231)

Implemented smart leverage defaults that remember user preferences per market:

- **Auto-Save on Order**: When user places an order, leverage is automatically saved for that market (per network)
- **Per-Market Memory**: Different assets remember different leverage (e.g., 6x for BTC, 10x for ETH, 3x for kPEPE)
- **Network-Specific**: Separate configurations for testnet and mainnet
- **Smart Priority**: Leverage selection follows: Navigation param > Existing position leverage (protocol constraint) > Saved config (user preference) > Default (3x)
- **Protocol Awareness**: Existing position logic unchanged - HyperLiquid still enforces new orders >= existing position leverage to prevent failures
- **Extensible Structure**: State structure allows future additions (e.g., size, TP/SL presets)

### 7. **WebSocket Cache Optimization**

Implemented cache-first pattern for position and order data to eliminate redundant API calls:

- **Cache-First Pattern**: `getPositions()` and `getOpenOrders()` methods now check WebSocket cache before making API calls
- **Real-Time Data**: Cache maintained by active WebSocket subscriptions provides fresher data than REST API
- **Performance Impact**: Reduces API load and improves response times across home screen, positions list, and order views
- **Smart Fallback**: Optional `skipCache` parameter available for scenarios requiring guaranteed fresh data from API
- **Architecture**: Subscription service maintains synchronized caches (`cachedPositions`, `cachedOrders`) updated in real-time via WebSocket events
- **Benefits All Features**: Optimization benefits every feature in this PR - home carousels, batch operations, and navigation all leverage instant cached data

### Technical Improvements

- **No Breaking Changes**: Maintained all existing routes and component APIs
- **Component Reusability**: Shared `PerpsCloseSummary` eliminates code duplication across 3 close position UIs
- **Performance**: Parallel execution for batch operations with partial success handling
- **Type Safety**: Complete TypeScript types for all new hooks, components, and controller methods
- **Consistent Patterns**: All hooks use object parameter pattern for extensibility
- **Route Naming**: Renamed perps routes for clarity: `TRADING_VIEW` → `PERPS_TAB` (redirects to wallet tab), `MARKETS` → `PERPS_HOME` (home screen), `TRENDING` → `MARKET_LIST` (full list). Screen names preserved for navigation state compatibility.
- **No Pull-to-Refresh**: Intentionally omitted from PerpsHomeView because 95% of data (positions, orders, activity) auto-updates via WebSocket (see Section 7). Only markets data uses REST API and changes infrequently. This prevents false expectations where users pull-to-refresh but see no visible change since data is already live.

## **Changelog**

CHANGELOG entry: Added Perps home screen with positions/orders carousels, trending markets list, recent activity, market sorting/filtering, and batch close all positions/cancel all orders functionality

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/TAT-1538
Fixes: https://consensyssoftware.atlassian.net/browse/TAT-1539
Fixes: https://consensyssoftware.atlassian.net/browse/TAT-1049
Fixes: https://consensyssoftware.atlassian.net/browse/TAT-1815
Fixes: https://consensyssoftware.atlassian.net/browse/TAT-1252
Fixes: https://consensyssoftware.atlassian.net/browse/TAT-1231

## **Manual testing steps**

```gherkin
Feature: Perps Home Screen v2

  Scenario: User views new home screen layout
    Given user is on Perps tab
    And user has at least 1 open position
    And user has at least 1 open order

    When user navigates to Perps home screen
    Then user sees positions carousel with position cards
    And user sees orders carousel with order cards
    And user sees trending markets list with top 5 markets
    And user sees recent activity list with last 3 trades
    And user sees "Learn more about perps" component

  Scenario: User navigates from home carousels
    Given user is on Perps home screen
    And positions carousel shows 3 positions
    And orders carousel shows 2 orders
    And trending markets shows 5 markets

    When user taps "View all" on positions carousel
    Then user is redirected to full positions list

    When user taps "View all" on orders carousel
    Then user is redirected to transactions view

    When user taps "See all" on trending markets
    Then user is redirected to market list view with sorting options

  Scenario: User sorts markets by different criteria
    Given user is on market list view (from "See all")
    And markets are displayed with default sort (Volume, descending)

    When user taps sort field dropdown
    Then user sees bottom sheet with options: Volume, Price Change, Market Cap, Funding Rate, New Markets

    When user selects "Price Change"
    And user taps sort direction toggle
    Then markets are sorted by price change in ascending order
    And sort chips reflect current selection

  Scenario: User searches for specific market
    Given user is on market list view
    And user sees search bar at top

    When user types "BTC" in search bar
    Then market list filters to show only BTC-related markets
    And search continues to work with active sort settings

  Scenario: User closes all positions at once
    Given user is on Perps home screen
    And user has 3 open positions
    And "Close All" button is visible

    When user taps "Close All" button
    Then bottom sheet appears showing:
      - Total margin (sum of all positions)
      - Total unrealized P&L (aggregated)
      - Total fees (closing fees for all positions)
      - Amount user will receive (margin + P&L - fees)
      - Estimated points (if rewards enabled)
      - Fee discount indicator (if applicable)

    When user taps "Close All" confirm button
    Then all 3 positions are closed at market price
    And success toast shows "Successfully closed 3 positions"
    And user is returned to home screen
    And positions carousel is empty or shows remaining positions

  Scenario: User closes all positions with partial failure
    Given user has 2 open positions
    And one position will fail to close (simulated network error)

    When user taps "Close All" and confirms
    Then 1 position closes successfully
    And 1 position fails to close
    And partial success toast shows "Successfully closed 1 of 2 positions"
    And failed position remains in positions carousel

  Scenario: User cancels all orders at once
    Given user is on Perps home screen
    And user has 4 open orders
    And "Cancel All" button is visible

    When user taps "Cancel All" button
    Then confirmation modal appears showing order count

    When user taps "Cancel All" confirm button
    Then all 4 orders are cancelled
    And success toast shows "Successfully cancelled 4 orders"
    And orders carousel is empty or shows remaining orders

  Scenario: User closes single position using shared summary component
    Given user is on single position close view
    And position has $1000 margin and +$150 P&L

    When user views the summary section
    Then user sees identical layout and formatting as close all summary:
      - Margin: $1,150 (includes P&L: +$150)
      - Fees: -$5.75 (with discount indicator if applicable)
      - You'll receive: $1,144.25
      - Estimated points (if rewards enabled)

    When user taps info icons
    Then same tooltips appear as in close all flow

  Scenario: User accesses learn more content
    Given user is on Perps home screen
    And "Learn more about perps" component is visible at bottom

    When user taps "Learn More" button
    Then external browser opens to MetaMask perps educational content

  Scenario: User's leverage preference is saved per market
    Given user has never traded BTC before
    And user has never traded ETH before
    And user is on trade order view for BTC

    When user selects leverage 6x
    And user places order successfully
    Then leverage 6x is saved for BTC on current network

    When user navigates to ETH trade order view
    Then leverage defaults to 3x (not 6x from BTC)

    When user selects leverage 10x for ETH
    And user places order successfully
    Then leverage 10x is saved for ETH on current network

    When user closes app and reopens
    And user navigates to BTC trade order view
    Then leverage pre-selects to 6x (saved preference)

    When user navigates to ETH trade order view
    Then leverage pre-selects to 10x (saved preference)

  Scenario: Saved leverage works across networks independently
    Given user is on testnet
    And user has saved leverage 5x for BTC on testnet

    When user places BTC order at 5x
    And user switches to mainnet
    And user navigates to BTC trade order view
    Then leverage defaults to 3x (mainnet has no saved config yet)

    When user places BTC order at 8x on mainnet
    And user switches back to testnet
    And user navigates to BTC trade order view
    Then leverage pre-selects to 5x (testnet config unchanged)

  Scenario: Existing position leverage takes priority over saved preference
    Given user has saved leverage 3x for BTC
    And user has open BTC position with leverage 10x
    And user navigates to BTC trade order view

    When order form loads
    Then leverage is set to 10x (existing position, not saved 3x)
    And this prevents order failure due to protocol constraint

  Scenario: Navigation param has highest priority for leverage
    Given user has saved leverage 6x for ETH
    And user has no open ETH position
    And user navigates to ETH order view with leverage=12 in params

    When order form loads
    Then leverage is set to 12x (navigation param)
    And saved preference 6x is not applied

  Scenario: Leverage priority chain validation
    Given user has saved leverage 5x for BTC
    And user has open BTC position with leverage 8x
    And user taps "Add to position" from position details (no nav param)

    When order form loads
    Then leverage is 8x (existing position wins over saved config)

    Given user has saved leverage 5x for ETH
    And user has no open ETH position
    And user navigates to ETH order view normally

    When order form loads
    Then leverage is 5x (saved config wins over default)

    Given user has never traded kPEPE
    And user has no open kPEPE position
    And user navigates to kPEPE order view

    When order form loads
    Then leverage is 3x (default fallback)
```

## **Screenshots/Recordings**

### **Before**

- Previous market list view showed only markets
- No home dashboard or overview
- No way to close all positions at once
- No market sorting or filtering
- Duplicate JSX for close position summaries

### **After**

- Comprehensive home screen with carousels and activity
- One-tap access to positions, orders, markets, and activity
- Close All and Cancel All buttons for batch operations
- Full market sorting and search functionality
- Shared `PerpsCloseSummary` component across all close flows
- Consistent UX patterns throughout

https://github.com/user-attachments/assets/66777d51-4f5a-436a-99c6-c375693e830e


## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I've included tests if applicable
- [x] I've documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I've applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

---



<!-- CURSOR_SUMMARY -->
> [!NOTE]
> Adds a new Perps Home dashboard, market sorting/search/watchlist, and batch Close All/Cancel All with controller + stream optimizations.
> 
> - **UI/UX**:
>   - **Home Screen v2**: Positions/orders carousels, trending markets, recent activity; new headers/components (HomeHeader, MarketListHeader, MarketList, RowSkeleton, NavigationCard, Section blocks).
>   - **Market List**: Sorting (volume/price change/open interest/funding), search, watchlist toggle; bottom-sheet sort picker.
>   - **Batch Actions**: Close All Positions modal (shared `PerpsCloseSummary`), Cancel All Orders modal; updated order/position views.
> - **Controllers/Provider**:
>   - **Batch APIs**: `PerpsController.closePositions()`/`cancelOrders()`; HyperLiquidProvider batch `closePositions`/`cancelOrders` + TWAP/transfer enhancements.
>   - **State**: Watchlist (per-network), saved trade config (per-market leverage), market sort preference; new selectors.
> - **Streams/Performance**:
>   - WebSocket cache-first for `getPositions`/`getOpenOrders`; stream pause/resume; cache init flags; market data prefetch tweaks.
> - **Routing**:
>   - Renames: `PERPS_TAB` (redirect), `PERPS_HOME` (dashboard), `MARKET_LIST` (full list); adds modal routes for batch actions.
> - **Utils/Data**:
>   - Market transform adds HIP-3 asset market typing and open interest; new market sorting utilities; format tweaks; search/sorting hooks.
> - **Tests**:
>   - Comprehensive unit tests for new hooks/components/controllers and batch flows.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 95b8c0298a95535426daedbbfd3303f4541c9c13. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

abretonc7s committed:
cleanup

abretonc7s committed:
docs: align doc format

abretonc7s committed:
feat: align error handling with perps controller

abretonc7s committed:
Merge branch 'main' into feat/perps/optimize-connection

abretonc7s committed:
improved reconnection flow

abretonc7s committed:
restore test

abretonc7s committed:
fix: unit test coverage

abretonc7s committed:
fix: pr comments

nickewansmith committed:
Merge branch 'main' into feat/perps/optimize-connection

abretonc7s committed:
Merge branch 'main' into feat/perps/optimize-connection

abretonc7s committed:
Merge branch 'feat/perps/optimize-connection' into feat/perps/optimize-connection2

abretonc7s committed:
tsc

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/perps/optimize-connection2

abretonc7s committed:
add coverage

abretonc7s committed:
feat: add sentry error identifier

abretonc7s committed:
feat: restore test deletion

abretonc7s committed:
fix: unit tests

abretonc7s committed:
feat: update sentry doc

abretonc7s committed:
:Merge remote-tracking branch 'origin/main' into feat/perps/optimize-connection2

abretonc7s committed:
fix: pr comments

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/perps/optimize-connection2

abretonc7s committed:
fix: missing reward discount when closing position

abretonc7s committed:
fix: unit tests coverage

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/perps/optimize-connection2

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/perps/optimize-connection2

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/perps/optimize-connection2

abretonc7s committed:
fix: pr comments

abretonc7s committed:
fix: critical type safety issue in tracing system

abretonc7s committed:
fix: unit tests

abretonc7s committed:
cleanup

abretonc7s committed:
fix: revert uistartup change

abretonc7s committed:
fix: unit test

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into fix/sentry-type

abretonc7s committed:
fix: unit tests

abretonc7s committed:
fix; discount emission on perps

abretonc7s committed:
fix: unit tests

abretonc7s committed:
fix: pr comments

abretonc7s committed:
cleanup

abretonc7s committed:
Merge remote-tracking branch 'origin/fix/sentry-type' into feat/perps/optimize-connection2

abretonc7s committed:
wip

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/perps/optimize-connection2

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/perps/optimize-connection2

abretonc7s committed:
cleanup

abretonc7s committed:
cleanup

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/perps/optimize-connection2

abretonc7s committed:
feat(perps): migrate to latest hyperliquid sdk

abretonc7s committed:
Merge branch 'main' into feat/perps/optimize-connection2

abretonc7s committed:
Merge branch 'feat/perps/optimize-connection2' into feat/perps/hypersdk-update

abretonc7s committed:
feat: remove ws package

abretonc7s committed:
cleanup

abretonc7s committed:
cleanup

abretonc7s committed:
cleanup

abretonc7s committed:
Merge branch 'main' into feat/perps/optimize-connection2

abretonc7s committed:
Merge branch 'feat/perps/optimize-connection2' into feat/perps/hypersdk-update

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/perps/hypersdk-update

abretonc7s committed:
fix: merge conflicts

abretonc7s committed:
cleanup

abretonc7s committed:
Merge branch 'main' into feat/perps/hypersdk-update

abretonc7s committed:
merge main into branch

abretonc7s committed:
fix: restore main behavior on formatting auto merged

abretonc7s committed:
cleanup + sonarcloud

abretonc7s committed:
feat(perps): HIP-3 support

abretonc7s committed:
Merge branch 'main' into feat/perps/hypersdk-update

abretonc7s committed:
feat: baseline for equity perps

abretonc7s committed:
feat: hyperliquid docs embedded for llm

abretonc7s committed:
feat: ui changes for hip3

abretonc7s committed:
working state with tpsl live

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/hip3

abretonc7s committed:
cleanup

abretonc7s committed:
cleanup

abretonc7s committed:
cleanup

abretonc7s committed:
fix: unit tests

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/hip3

abretonc7s committed:
fix: bugbot

abretonc7s committed:
cleanup removing magic nums

abretonc7s committed:
Merge branch 'main' into perps/feat/hip3

abretonc7s committed:
fix: linting issues

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/hip3

abretonc7s committed:
Merge branch 'main' into perps/feat/hip3

abretonc7s committed:
feat: styling badges

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/hip3

abretonc7s committed:
cleanup

abretonc7s committed:
fix: bugbot comment

abretonc7s committed:
cleanup

abretonc7s committed:
cleanup

abretonc7s committed:
cleanup

abretonc7s committed:
cleanup

abretonc7s committed:
cleanup

abretonc7s committed:
feat: re-enable hip3 debug page for testing

abretonc7s committed:
feat: cleanup

abretonc7s committed:
cleanup

abretonc7s committed:
feat: feature flag improvement and controller init

abretonc7s committed:
cleanup

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/hip3

abretonc7s committed:
feat: pr commentS

abretonc7s committed:
fix: unit tests

abretonc7s committed:
feat: change default equityEnabled flag to true on dev only

abretonc7s committed:
fix: unit test

abretonc7s committed:
fix: unit tests

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/hip3

abretonc7s committed:
fix: sonarcloud issues

abretonc7s committed:
fix: linting

abretonc7s committed:
fix: sonarcloud

abretonc7s committed:
fix: sonarcloud

abretonc7s committed:
cleanup

abretonc7s committed:
cleanup

abretonc7s committed:
feat: cleanup

abretonc7s committed:
feat: tat-1784

abretonc7s committed:
cleanup

abretonc7s committed:
cleanup

abretonc7s committed:
coverage

abretonc7s committed:
coverage

abretonc7s committed:
feat: coverage

abretonc7s committed:
feat(perps): add home screen v2, market sorting, and batch operations

abretonc7s committed:
cleanup

abretonc7s committed:
cleanup

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/newhome

abretonc7s committed:
cleanup

abretonc7s committed:
fix: unit test

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/hip3

abretonc7s committed:
pr comment

abretonc7s committed:
feat: add error handling to perpsdex

abretonc7s committed:
fix: overfund dex to allow acreating new position over existing

abretonc7s committed:
merge main into branch

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/newhome

abretonc7s committed:
fix: cleanup

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/hip3

abretonc7s committed:
testing auto transfer

abretonc7s committed:
feat: hip3 abstraction support

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/hip3

abretonc7s committed:
fix: tpsl matching

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/hip3

abretonc7s committed:
merge main into branch

abretonc7s committed:
feat: utility funciton to determine maker status

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/newhome

abretonc7s committed:
merge hip3 into branch

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/newhome

abretonc7s committed:
feat(sentry): add searchable tags and context to Logger.error

abretonc7s committed:
feat: add unit tests

abretonc7s committed:
Merge branch 'main' into feat/sentry-tags

abretonc7s committed:
feat: closing all positions

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/newhome

abretonc7s committed:
fix: unit testS

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/sentry-tags

abretonc7s committed:
fix: coverage

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/newhome

abretonc7s committed:
feat: refactor for unit test coverage

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/newhome

abretonc7s committed:
Merge branch 'main' into feat/sentry-tags

abretonc7s committed:
fix: unit tests

abretonc7s committed:
feat: handle legazy format

abretonc7s committed:
fix(perps): improve error handling and navigation safety

- Add Sentry logging for stream pause/resume failures
- Fix race condition using Promise-based guard in cancelOrders
- Add navigation.canGoBack() checks with Markets fallback
- Clone enabledDexs array to prevent external mutations
- Update test mocks to include canGoBack method

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/newhome

abretonc7s committed:
Merge branch 'main' into feat/sentry-tags

abretonc7s committed:
feat: unit test

abretonc7s committed:
merge main into branch

abretonc7s committed:
Merge branch 'main' into feat/sentry-tags

abretonc7s committed:
Merge branch 'feat/sentry-tags' into perps/feat/newhome

abretonc7s committed:
fix: resolve duplicate symbol display and invalid concurrency test

abretonc7s committed:
fix: unit test and linting

abretonc7s committed:
feat(perps): add watchlist feature

abretonc7s committed:
feat: watchlist support and unit tests

abretonc7s committed:
fix: unit tests

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/newhome

abretonc7s committed:
fix: unit tests

abretonc7s committed:
feat: compute points based on currentprice not entryprice

abretonc7s committed:
feat: styling

abretonc7s committed:
feat: perps tab card navigation on everytabs

abretonc7s committed:
fix: bugbot

abretonc7s committed:
fix: unit tests

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/newhome

abretonc7s committed:
feat: support leverage trading preferences

abretonc7s committed:
coverage

abretonc7s committed:
feat: coverage and disable pull to refresh

abretonc7s committed:
feat: add loading indicator while loading fees

abretonc7s committed:
fix: perps recent activity list

abretonc7s committed:
feat: trending sorting page

abretonc7s committed:
feat: persisting trending sort preferences

abretonc7s committed:
fix: unit tests

abretonc7s committed:
fix: merge main into branch

abretonc7s committed:
fix: unit test + linting

abretonc7s committed:
fix: coverage

abretonc7s committed:
merge main into branch

abretonc7s committed:
cleanup

abretonc7s committed:
styling

abretonc7s committed:
fix: unit tests

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/newhome

abretonc7s committed:
feat: improve caching access

abretonc7s committed:
feat: merge main into branch

abretonc7s committed:
feat: perps card nav under tab

abretonc7s committed:
feat; unit tests

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/newhome

abretonc7s committed:
docs: improvements

abretonc7s committed:
fix: linting

abretonc7s committed:
fix: unit test

abretonc7s committed:
feat: use e2e selectors for testids

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/newhome

abretonc7s committed:
fix: unit tests

abretonc7s committed:
shared styles to avoid duplication

abretonc7s committed:
feat: unit test

abretonc7s committed:
sonarcloud

abretonc7s committed:
sonarcloud

abretonc7s committed:
sonarcloud

abretonc7s committed:
cleanup

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/newhome

abretonc7s committed:
cleanup

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/newhome

gambinish commented:
Noticing that when using the search bar, that the keyboard pushes the wallet footer up, rather than rendering over it. I'm not sure if this was introduced here, but I don't remember this behaving this way before. We may want to double check with design here.

<img width="441" height="939" alt="Screenshot 2025-10-24 at 9 23 06 AM" src="https://github.com/user-attachments/assets/763cff8c-fe70-4b1b-8492-a49d908101d1" />

gambinish commented:
Can we introduces a loading skeleton here to prevent this slight jank that I see on this navigation? When I press `Available balance` I am navigated to the home screen, which initially empty of records `Loading...`, and then it re-renders once the records are loaded:

https://github.com/user-attachments/assets/c284cd6e-001e-4138-980c-b229c4305741

We do something similar on the Perps transaction view which smooths this out a bit

georgewrmarshall commented on app/components/UI/Perps/Views/PerpsHomeView/PerpsHomeView.tsx:
```diff
@@ -0,0 +1,457 @@
+        </View>
+      </Box>
+    </View>
+  );
```
Yeah what are the reasons for not using the component from the home page?

abretonc7s replied:
yes good point, I re-used what was done here but better with a component

abretonc7s commented on app/components/UI/Perps/controllers/PerpsController.ts:
```diff
@@ -1702,6 +2043,155 @@ export class PerpsController extends BaseController<
+
+        // Close all positions in parallel using Promise.allSettled
+        const results = await Promise.allSettled(
+          positionsToClose.map((position) =>
```
But we would still need to make multiple call per markets

abretonc7s commented on locales/languages/en.json:
```diff
@@ -416,9 +416,9 @@
-    "description_social_login_update" : "If you lose this password, you’ll lose access to your wallet on all devices. Store it somewhere safe,",
-    "description_social_login_update_ios" : "Use this for wallet recovery on all devices. MetaMask can’t reset it.",
-    "description_social_login_update_bold" : "MetaMask can't reset it.",
+    "description_social_login_update": "If you lose this password, you’ll lose access to your wallet on all devices. Store it somewhere safe,",
```
Probably but completely unrelated to this Pr, not sure why the changes got picked up during rebase

abretonc7s committed:
merge main into branch

abretonc7s committed:
feat: cleanup

abretonc7s committed:
feat: optimize batch cancel close orders positions

abretonc7s committed:
feat: improved loaders in home screen

abretonc7s committed:
feat: perps home view styling with market list

abretonc7s committed:
feat: simplify perps route for consistency

abretonc7s committed:
feat: cleanup ui component

abretonc7s committed:
feat: improved ui refactor with watchlist toggle

abretonc7s committed:
styling

abretonc7s committed:
fix: prevent invalid points computation

abretonc7s committed:
fix: unit tests

abretonc7s committed:
fix: linting

abretonc7s committed:
feat: navigation via hooks

abretonc7s committed:
fix: bugbot

abretonc7s committed:
fix: unit test coverage

abretonc7s committed:
feat: improved leverage validation

abretonc7s committed:
fix: unit tests

abretonc7s committed:
fix: unit tests

abretonc7s committed:
fix: unit tests

abretonc7s committed:
fix: unit test coverage

abretonc7s committed:
fix: unit tests

abretonc7s committed:
fix: unit tests

abretonc7s committed:
fix: coverage

abretonc7s committed:
fix: sonarcloud

abretonc7s committed:
feat: use crypto instead of perps

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/newhome

abretonc7s committed:
fix: bugbot

abretonc7s committed:
feat: hide bottom nav on marketlist

abretonc7s committed:
feat: remove watchlist from sort fields

gambinish commented:
The new scrollable tab view's behavior seems buggy to me. I can often end up in a state with the highlighted text and the underline do not match. And I can navigate to a tab by swiping without the change getting reflected in the tab view component

https://github.com/user-attachments/assets/a4525632-a406-4e70-b5e0-29b16597dc4e

https://github.com/user-attachments/assets/6a513e98-7a59-4b94-bed1-b3d404303a80

gambinish commented on app/components/UI/Bridge/components/TokenInputArea/TokenInputArea.test.tsx:
```diff
@@ -299,8 +299,8 @@ describe('getDisplayAmount', () => {
-    );
+    expect(
+      getDisplayAmount(amount, TokenInputAreaType.Destination, false),
+    ).toBe('123.45678');
```
This looks like another formatting diff

gambinish committed:
chore: Merge main, address conflicts

cortisiko committed:
test(predict): add cash out position e2e  (#21545)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

The purpose of this test is to increase predictions e2e coverage by
adding a cash out position test. Scenario 4 from Test Plan
## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the
CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing
description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and
accurately)
-->

CHANGELOG entry:

## **Related issues**

Fixes:

## **Manual testing steps**

```gherkin
Feature: my feature name

  Scenario: user [verb for user action]
    Given [describe expected initial app state]

    When user [verb for user action]
    Then [describe expected outcome]
```

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.


<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds an end-to-end cash-out flow test with comprehensive Polymarket
API/RPC/order-book mocks and new testIDs to enable stable UI automation.
> 
> - **E2E Tests**:
> - Add `predict-cash-out.spec.ts` covering cash-out flow (balances
update, activity entry, position removal) and related page objects.
> - Update `predict-api-down.spec.ts` to use new Polymarket mocks path.
> - **Mocks (Polymarket)**:
> - Introduce detailed mocks: event details (dynamic by `eventId`),
current/resolved positions (with optional winnings), order book per
`token_id`, activity, UpNL, and complete setup helpers.
> - Add post-action helpers: `POLYMARKET_POST_CASH_OUT_MOCKS`,
`POLYMARKET_FORCE_BALANCE_REFRESH_MOCKS`,
`POLYMARKET_REMOVE_CASHED_OUT_POSITION_MOCKS`, and post-claim mocks.
> - New constants for USDC balances (initial/post-claim/post-cash-out)
and addresses; track dynamic USDC balance in RPC mocks.
> - Remove legacy `polymarket-mocks.ts` and split into organized files.
> - **General Mocks**:
> - Expand price API defaults and add `polygon-rpc.com` default RPC
mock.
> - **Selectors & Pages**:
> - Add Predict Activity Details selectors and page; extend
ActivitiesView with predictions tab and helpers; refine predictions
selectors (stable position card IDs).
> - **App UI (testIDs)**:
> - Add `testID`s to `PredictActivityDetail` (back, title, amount,
container).
>   - Add `testID` to market details cash-out Button.
>   - Adjust position components’ `testID`s for stable selection.
> 
> <sup>Written by [Cursor
Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit
02b594feb62b7327329904d871dc6cde55edb7ac. This will update automatically
on new commits. Configure
[here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

gambinish committed:
Merge branch 'main' into perps/feat/newhome

tommasini committed:
chore: revert hardcoded fingerprint (#21617)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the
CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing
description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and
accurately)
-->

CHANGELOG entry:

## **Related issues**

Fixes:

## **Manual testing steps**

```gherkin
Feature: my feature name

  Scenario: user [verb for user action]
    Given [describe expected initial app state]

    When user [verb for user action]
    Then [describe expected outcome]
```

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Replace hardcoded fingerprint with `yarn fingerprint:generate` in
`.github/workflows/build-android-e2e.yml` to compute the cache key
dynamically.
> 
> <sup>Written by [Cursor
Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit
47f5d9a1da6406c6111a8f0d0736c848e40a7f1e. This will update automatically
on new commits. Configure
[here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

---------

Co-authored-by: metamaskbot <metamaskbot@users.noreply.github.com>

bsgrigorov committed:
fix(INFRA-3066): path based rebuilds (#21747)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

This PR fixes a critical bug in the GitHub Actions workflow that was
preventing native code changes from triggering E2E test rebuilds.

**Problem:** The negation pattern `!scripts/build.sh` in the paths
filter was breaking the entire filtering logic, causing the
dorny/paths-filter action to incorrectly treat all files (including
native Android and iOS code) as ignored. This meant that PRs with only
native code changes would not trigger app rebuilds, potentially running
E2E tests against stale builds.

**Solution:**
1. Moved `scripts/build.sh` from the `ignore` section (with problematic
negation) to the `shared` section
2. Added comprehensive documentation explaining the intended filter
logic
3. Ensured native code changes (`android/**`, `ios/**`) properly trigger
platform-specific rebuilds

This fix ensures the CI workflow correctly identifies when rebuilds are
needed while maintaining the intended optimization of avoiding
unnecessary rebuilds for documentation-only changes.

## **Changelog**

CHANGELOG entry: null

## **Related issues**

## **Manual testing steps**

1. Create a test PR with only Android native code changes (e.g., modify
a file in `android/app/src/`)
2. Verify the `needs_e2e_build` job outputs `android_changed: true` and
`builds: true`
3. Create a test PR with only iOS native code changes (e.g., modify a
file in `ios/`)
4. Verify the `needs_e2e_build` job outputs `ios_changed: true` and
`builds: true`
5. Create a test PR with only documentation changes (e.g., modify
`README.md`)
6. Verify the `needs_e2e_build` job outputs `builds: false` (no
unnecessary rebuilds)
7. Create a test PR that modifies `scripts/build.sh`
8. Verify it triggers rebuilds for both platforms (`android_changed:
true`, `ios_changed: true`)

## **Screenshots/Recordings**

### **Before**

Native code changes were incorrectly flagged as ignored:

### **After**

Native code changes now properly trigger the appropriate platform
builds.

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I've included tests if applicable
- [x] I've documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I've applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.


<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Refines path filters in `needs-e2e-build.yml` by moving
`scripts/build.sh` to `shared`, clarifying ignore rules, and adding
clear documentation to ensure correct Android/iOS rebuild triggers.
> 
> - **CI Workflow (`.github/workflows/needs-e2e-build.yml`)**:
>   - **Path filtering**:
> - Move `scripts/build.sh` from `ignore` to `shared` to trigger
rebuilds on changes.
> - Keep native paths `android/**` and `ios/**` as platform-specific
triggers; group common/dependency files under `shared`.
> - Maintain `ignore` for docs/workflow and utility scripts; refine
patterns and comments for clarity.
>     - Preserve `catch_all` fallback and output listing configuration.
> - **Documentation**: Add detailed comments explaining filter intent
and key points.
> 
> <sup>Written by [Cursor
Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit
ddc5d0b0b907563ded27b8f86f13d4dd575fe479. This will update automatically
on new commits. Configure
[here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

abretonc7s commented:
> From your PR description:
> 
> > Positions Carousel: Horizontal scrollable list of open positions with quick access to position details
> > Orders Carousel: Horizontal scrollable list of open limit orders with quick navigation
> 
> Can you provide a demo of this behavior? I have been trying to get to this to test and cannot seem to find where it was implemented.

Ok so I haven't updated the description, but basically the design has changed while implementing and this is not a carousel anymore...

abretonc7s committed:
cleanup

abretonc7s committed:
Merge branch 'main' into perps/feat/newhome

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/newhome

abretonc7s committed:
reset to main

abretonc7s committed:
reset to main

abretonc7s committed:
reset to main

abretonc7s committed:
reset to main

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into perps/feat/newhome

gambinish commented on app/components/UI/Perps/styles/sharedStyles.ts:
```diff
@@ -0,0 +1,26 @@
+    emptyText: {
+      paddingHorizontal: 16,
+    },
+  });
```
That's annoying. I'll resolve this now but something to consider refactoring down the line

gambinish committed:
chore: Merge main, address conflicts
